fiat-roles/src/main/java/com/netflix/spinnaker/fiat/permissions/RedisPermissionsRepository.java

```diff
@@ -76,43 +82,120 @@
+
   private final LoadingCache<String, UserPermission> unrestrictedPermission =
       Caffeine.newBuilder()
           .expireAfterAccess(Duration.ofSeconds(10))
           .build(k -> reloadUnrestricted());
 
   private final String prefix;
 
-  public RedisPermissionsRepository(
+  RedisPermissionsRepository(
+      Clock clock,
       ObjectMapper objectMapper,
       RedisClientDelegate redisClientDelegate,
       List<Resource> resources,
-      String prefix) {
+      RedisPermissionRepositoryConfigProps configProps,
+      RetryRegistry retryRegistry) {
+    this.clock = clock;
     this.objectMapper = objectMapper;
     this.redisClientDelegate = redisClientDelegate;
-    this.prefix = prefix;
+    this.configProps = configProps;
+    this.prefix = configProps.getPrefix();
     this.resources = resources;
+    this.retryRegistry = retryRegistry;
+  }
+
+  public RedisPermissionsRepository(
+      ObjectMapper objectMapper,
+      RedisClientDelegate redisClientDelegate,
+      List<Resource> resources,
+      RedisPermissionRepositoryConfigProps configProps,
+      RetryRegistry retryRegistry) {
+    this(
+        Clock.systemUTC(),
+        objectMapper,
+        redisClientDelegate,
+        resources,
+        configProps,
+        retryRegistry);
   }
 
   private UserPermission reloadUnrestricted() {
-    return getFromRedis(UNRESTRICTED).orElseThrow(NoSuchElementException::new);
+    return getFromRedis(UNRESTRICTED)
+        .orElseThrow(() -> new PermissionRepositoryException("Failed to read unrestricted user"));
   }
 
   private UserPermission getUnrestrictedUserPermission() {
     String serverLastModified =
-        redisClientDelegate.withCommandsClient(
-            c -> {
-              return c.get(unrestrictedLastModifiedKey());
-            });
+        redisRead(
+            new TimeoutContext(
+                "checkLastModified",
+                clock,
+                configProps.getRepository().getCheckLastModifiedTimeout()),
+            c -> c.get(unrestrictedLastModifiedKey()));
     if (serverLastModified == null) {
       serverLastModified = NO_LAST_MODIFIED;
     }
     return unrestrictedPermission.get(serverLastModified);
   }
 
+  private static class TimeoutContext {
+    private final String name;
+    private final long timeoutAtMs;
+    private final Clock clock;
+
+    public TimeoutContext(String name, Clock clock, Duration timeout) {
+      this(name, clock, timeout, Instant.now(clock));
+    }
+
+    public TimeoutContext(String name, Clock clock, Duration timeout, Instant startTime) {
+      this.name = name;
+      this.timeoutAtMs = startTime.plus(timeout).toEpochMilli();
+      this.clock = clock;
+    }
+
+    boolean isTimeout() {
+      return clock.millis() > timeoutAtMs;
+    }
+
+    String getName() {
+      return name;
+    }
+  }
+
+  private <T> T redisRead(TimeoutContext timeoutContext, Function<JedisCommands, T> fn) {
+    return retryRegistry
```

Comment:
need to do the retry this way as opposed to annotation based because the AOP stuff doesn't kick in on an internal method call within a class